price-feeder/config/currency_provider_tracker.go

```diff
@@ -0,0 +1,175 @@
+		return nil, err
+	}
+
+	currencyProviderTracker.setCurrencyProviderMin()
+
+	go currencyProviderTracker.trackCurrencyProviders(ctx)
+
+	return currencyProviderTracker, nil
+}
+
+func (t *CurrencyProviderTracker) getCurrencyProviders() map[string][]string {
+	return t.CurrencyProviders
+}
+
+func (t *CurrencyProviderTracker) getCurrencyProviderMin(base string) int {
+	return t.CurrencyProviderMin[base]
+}
+
+func (t *CurrencyProviderTracker) logCurrencyProviders() {
+	for currency, providers := range t.CurrencyProviders {
+		t.logger.Info().Msg(fmt.Sprintf("Providers supporting %s: %v", currency, providers))
+	}
+}
+
+// setCoinIDSymbolMap gets list of assets on coingecko to cross reference coin symbol to id.
+func (t *CurrencyProviderTracker) setCoinIDSymbolMap() error {
+	resp, err := http.Get(coinGeckoRestHost + coinGeckoRestPath + coinGeckoListEndpoint)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	var listResponse []CoinGeckoCoinList
+	if err := json.NewDecoder(resp.Body).Decode(&listResponse); err != nil {
+		return err
+	}
+
+	for _, coin := range listResponse {
+		t.coinIDSymbolMap[coin.Symbol] = coin.ID
+	}
+
+	return nil
+}
+
+// setCurrencyProviders queries CoinGecko's tickers endpoint to get all the exchanges that
+// support each price feeder currency pair and store it in the CurrencyProviders map
+func (t *CurrencyProviderTracker) setCurrencyProviders() error {
+	for _, pair := range t.pairs {
+		pairBaseID := t.coinIDSymbolMap[strings.ToLower(pair.Base)]
+		resp, err := http.Get(coinGeckoRestHost + coinGeckoRestPath + pairBaseID + "/" + coinGeckoTickersEndpoint)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		var tickerResponse CoinGeckoCoinTickerResponse
+		if err = json.NewDecoder(resp.Body).Decode(&tickerResponse); err != nil {
+			return err
+		}
+
+		for _, ticker := range tickerResponse.Tickers {
+			if ticker.Target == pair.Quote {
+				t.CurrencyProviders[pair.Base] = append(t.CurrencyProviders[pair.Base], ticker.Market.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// setCurrencyProviderMin will set the minimum amount of providers for each currency
+// to the amount of exchanges that support them if its less than 3. Otherwise it is
+// set to 3 providers.
+func (t *CurrencyProviderTracker) setCurrencyProviderMin() {
+	for base, exchanges := range t.getCurrencyProviders() {
+		if len(exchanges) < 3 {
+			t.CurrencyProviderMin[base] = len(exchanges)
+		} else {
+			t.CurrencyProviderMin[base] = 3
+		}
+	}
+}
+
+// trackCurrencyProviders resets CurrencyProviders map and logs out supported
+// exchanges for each currency every 24 hours
+func (t *CurrencyProviderTracker) trackCurrencyProviders(ctx context.Context) {
+	t.logCurrencyProviders()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(trackingPeriod):
+			if err := t.setCurrencyProviders(); err != nil {
+				t.logger.Error().Err(err).Msg("failed to set available providers for currencies")
```

Comment:
This is great! Question, what happens if on startup the validator is unable to reach the coingecko API, but they have 3 providers per asset? Will it not start up?

Reply:
That's a good point. Unless you prefer an alternate approach, I'm gonna make it so that if the request fails it will just set the minimum providers to 3 by default and just log that the request failed for that currency.

Reply:
@rbajollari what if coingecko is down for an extended period of time, say a week? We've then got a bottleneck for validators' votes. Maybe we need an override in the config?

Wdyt @zarazan ?

Reply:
If coingecko is down I think you should be able to start the price-feeder and everything should work normally except a warning that it was unable to do currency checks. That means if we allow a single provider when there is only one available that should still be allowed if coingecko is down.

Reply:
A config override is a decent option. It should probably just bypass the validation completely if you use the override.

Reply:
Override > defaulting to allowing any amount of providers imo, at least at startup - just because we might never hear about it if validators don't have to set something manually

Reply:
So if the override config just doesn't enforce the provider minimum when set, does it take precedence over the coin gecko check or only take effect when the coin gecko request fails? @adamewozniak

Reply:
Good question, I think it should only take effect when the coingecko request fails @rbajollari

Reply:
@adamewozniak Implemented that as `provider_min_override` config. Lmk what you think about it.

Reply:
Perfect!